File: dotnet/src/Datastructure/Otterop/Datastructure/LinkedListIterator.cs
namespace Otterop.Datastructure
{
    class LinkedListIterator<T> : Otterop.Lang.OOPIterator<T>
    {
        private LinkedListNode<T> current;
        internal LinkedListIterator(LinkedList<T> linkedList)
        {
            this.current = linkedList.First();
        }

        public bool HasNext()
        {
            return this.current != null;
        }

        public T Next()
        {
            LinkedListNode<T> ret = this.current;
            this.current = this.current.Next();
            return ret.Value();
        }

    }

}

File: java/datastructure/src/main/java/otterop/datastructure/LinkedListIterator.java
package otterop.datastructure;

import otterop.lang.OOPIterator;

class LinkedListIterator<T> implements OOPIterator<T> {

    private LinkedListNode<T> current;

    LinkedListIterator(LinkedList<T> linkedList) {
        this.current = linkedList.first();
    }
    
    public boolean hasNext() {
        return this.current != null;
    }

    public T next() {
        LinkedListNode<T> ret = this.current;
        this.current = this.current.next();
        return ret.value();
    }
}
